### PR TITLE
Fixes #1219: Add UTC time in TimeAndDateService

### LINKED ISSUE #1219: Add UTC time in TimeAndDateService
- Issue type: Feature request

At present, TimeAndDateService only outputs local time and date. A Feature to output UTC time and time difference between the two shall also be added.

### PR DESCRIPTION
**Pull request by vibhcool**
This PR :-

1) Fixes Issue #1219
2) Minor fixes in LocationWiseTimeService.java , I found while reading code.

I have:
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only strictly only one commit per issue.

### For the reviewers
I have:
- [ ] Reviewed this pull request by an authorized contributor.
- [ ] The reviewer is assigned to the pull request.

### CODE REVIEW

**Comment by vibhcool** [comment]
I am not sure, about codacy issue.
Please review 😄

**Comment by vibhcool** [comment]
@SKrPl @singhpratyush @djmgit @Achint08 @SarishiNoHara @hemantjadon @kavithaenair

File: src/org/loklak/api/search/TimeAndDateService.java [hunk 1]
```diff
@@ -72,4 +81,34 @@ public static SusiThought timeAndDate() {
...
+
+        SimpleDateFormat dateFormatGmt = new SimpleDateFormat("yyyy-MMM-dd HH:mm:ss");
+        dateFormatGmt.setTimeZone(TimeZone.getTimeZone("GMT"));
+
+        //Local time zone   
+        SimpleDateFormat dateFormatLocal = new SimpleDateFormat("yyyy-MMM-dd HH:mm:ss");
+
+        try {
+        //Time in GMT
+        date = dateFormatLocal.parse(dateFormatGmt.format(date));
```
Comment (singhpratyush) [inline src/org/loklak/api/search/TimeAndDateService.java]: The codacy issue is because you reassigned this object here. This check ensures that you do not accidentally modify an object in a function.

Using a new variable like `newDate` would fix the issue.

File: src/org/loklak/api/search/TimeAndDateService.java [hunk 2]
```diff
@@ -72,4 +81,34 @@ public static SusiThought timeAndDate() {
...
+        }
+        return date;
+    }
+
+    private static String timeDiff(Date timeLocal) {
+    
+        long diff = timeLocal.getTime() - timeToUTC(timeLocal).getTime();
+        long diffMin = diff / (60 * 1000) % 60;
+        long diffHrs = diff / (60 * 60 * 1000) % 24;
+        String diffTime = String.valueOf(diffHrs) + ":" + String.valueOf(diffMin).replace("-","");
```
Comment (singhpratyush) [inline src/org/loklak/api/search/TimeAndDateService.java]: Will this be returning strings like `+05:30` and not `5:30`? Do you think that it would be better to use `TimeZone#getOffset` instead of this?

**Comment by kavithaenair** [comment]
@hemantjadon the failure of Travis build is not related to this PR. The issue #1215 is already created here.

**Comment by hemantjadon** [comment]
@kavithaenair Okay cool 👍

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 2 above)
Comment (Achint08) [inline src/org/loklak/api/search/TimeAndDateService.java]: Yes.TimeZone#getOffset would be a good step here instead.

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 1 above)
Comment (vibhcool) [inline src/org/loklak/api/search/TimeAndDateService.java]: I meant we shall keep the change IMO 😅

File: src/org/loklak/api/search/LocationWiseTimeService.java [hunk 3]
```diff
@@ -21,15 +21,15 @@
...
-
-import org.json.JSONArray;
-import org.json.JSONObject;
 import org.jsoup.Jsoup;
 import org.jsoup.nodes.Document;
 import org.jsoup.nodes.Element;
 import org.jsoup.select.Elements;
+
+import org.json.JSONArray;
+import org.json.JSONObject;
```
Comment (kavithaenair) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: Any reason behind moving these lines?

Comment (singhpratyush) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: The `org.json` imports should be before `org.jsoup` imports.

Comment (kavithaenair) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: Yes @singhpratyush

File: src/org/loklak/api/search/TimeAndDateService.java [hunk 4]
```diff
@@ -62,8 +65,14 @@ public static SusiThought timeAndDate() {
...
 		
 		Date time_and_date = new Date();
-		timeAndDate.put("time_and_date", time_and_date.toString());
-		
+        String time_and_date_UTC = timeToUTC(time_and_date).toString().replaceAll(
+                time_and_date.toString().substring(20,23),
+                "UTC"
+        );
+
+        timeAndDate.put("local_time_and_date", time_and_date.toString());
```
Comment (SKrPl) [inline src/org/loklak/api/search/TimeAndDateService.java]: Please use uniform indentation, here spaces are used and in other lines tabs are used.

Comment (vibhcool) [inline src/org/loklak/api/search/TimeAndDateService.java]: @SKrPl , I have replaced all tabs by 4 spaces

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 2 above)
Comment (vibhcool) [inline src/org/loklak/api/search/TimeAndDateService.java]: @singhpratyush TimeZone.getOffset() and Date.getTime() gives same output, since we are dealing date object, I used .getTime()

File: src/org/loklak/api/search/LocationWiseTimeService.java (hunk 3 above)
Comment (vibhcool) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: @kavithaenair @singhpratyush , there a rule to keep imports from project and from other projects separate, isn't it?
`org.json` belongs to loklak server project and `org.jsoup` is other project's import

Comment (singhpratyush) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: It is about following the lexical order.

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 4 above)
Comment (singhpratyush) [inline src/org/loklak/api/search/TimeAndDateService.java]: @vibhcool: Make your IDE use the `.editorconfig` file introduced recently. You will not face such errors.

**Comment by vibhcool** [comment]
@kavithaenair @singhpratyush , fixed imports lexical order

**Comment by mariobehling** [comment]
Travis is still failing. Please ensure travis is not failing on new commits. This is the responsibility of developers even if travis failures existed beforehand.

File: src/org/loklak/api/search/TimeAndDateService.java [hunk 5]
```diff
@@ -35,41 +38,81 @@
...
+
+        SimpleDateFormat dateFormatGmt = new SimpleDateFormat("yyyy-MMM-dd HH:mm:ss");
+        dateFormatGmt.setTimeZone(TimeZone.getTimeZone("GMT"));
+
+        //Local time zone   
+        SimpleDateFormat dateFormatLocal = new SimpleDateFormat("yyyy-MMM-dd HH:mm:ss");
+
+        try {
+        //Time in GMT
+        date = dateFormatLocal.parse(dateFormatGmt.format(date));
```
Comment (singhpratyush) [inline src/org/loklak/api/search/TimeAndDateService.java]: Please fix the indentation here.

File: src/org/loklak/api/search/LocationWiseTimeService.java [hunk 6]
```diff
@@ -21,15 +21,14 @@
...
 import org.json.JSONArray;
 import org.json.JSONObject;
 import org.jsoup.Jsoup;
 import org.jsoup.nodes.Document;
 import org.jsoup.nodes.Element;
 import org.jsoup.select.Elements;
+import org.loklak.data.DAO;
 import org.loklak.server.APIException;
 import org.loklak.server.APIHandler;
 import org.loklak.server.AbstractAPIHandler;
```
Comment (singhpratyush) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: Should come before `org.loklak.server.APIException`.

Comment (vibhcool) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: fixed this :)

File: src/org/loklak/api/search/LocationWiseTimeService.java [hunk 7]
```diff
@@ -59,12 +58,16 @@ public JSONObject getDefaultPermissions(BaseUserRole baseUserRole) {
...
+    @Override
+    public JSONObject serviceImpl(
+            Query call,
+            HttpServletResponse response,
+            Authorization rights,
+            JSONObjectWithDefault permissions
+    ) throws APIException {
+        String query = call.get("query", "");
+        return locationWiseTime(query);
+    }
```
Comment (singhpratyush) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: The earlier state was good. Why this change?

Comment (vibhcool) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: earlier it didn't exceed line length, but didn't seem to be proper.

Comment (singhpratyush) [inline src/org/loklak/api/search/LocationWiseTimeService.java]: Maybe it's about the varying taste of code styling. But since it didn't violate any rules, I think it shouldn't be changed.

**Comment by kavithaenair** [comment]
@vibhcool check the travis build.

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 2 above)
Comment (vibhcool) [inline src/org/loklak/api/search/TimeAndDateService.java]: this string will be returning 5:30

Comment (singhpratyush) [inline src/org/loklak/api/search/TimeAndDateService.java]: And how will it handle negative timezones?

Comment (vibhcool) [inline src/org/loklak/api/search/TimeAndDateService.java]: negative timezones will appear like -5:30

**Comment by vibhcool** [comment]
@kavithaenair , rebased, error wasn't related to PR.

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 2 above)
Comment (singhpratyush) [inline src/org/loklak/api/search/TimeAndDateService.java]: Oh yes. Was looking at an outdated diff. Would it be better to pad zeros and sign to make it `+05:30` instead of `5:30`?

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 4 above)
Comment (vibhcool) [inline src/org/loklak/api/search/TimeAndDateService.java]: made the suggested change

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 2 above)
Comment (vibhcool) [inline src/org/loklak/api/search/TimeAndDateService.java]: made the changes :)

File: src/org/loklak/api/search/TimeAndDateService.java [hunk 8]
```diff
@@ -35,41 +38,82 @@
...
+
+        SimpleDateFormat dateFormatGmt = new SimpleDateFormat("yyyy-MMM-dd HH:mm:ss");
+        dateFormatGmt.setTimeZone(TimeZone.getTimeZone("GMT"));
+
+        //Local time zone   
+        SimpleDateFormat dateFormatLocal = new SimpleDateFormat("yyyy-MMM-dd HH:mm:ss");
+
+        try {
+           //Time in GMT
+           date = dateFormatLocal.parse(dateFormatGmt.format(date));
```
Comment (SarishiNoHara) [inline src/org/loklak/api/search/TimeAndDateService.java]: The question is that you reassign param `date` in the method. To avoid this, you can assign date to a local variable and then return it. see [https://stackoverflow.com/a/13507090/5767950](https://stackoverflow.com/a/13507090/5767950)

**Comment by vibhcool** [comment]
@SarishiNoHara @singhpratyush , fixed the codacy issue, you suggested was important :)
please re-review

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 8 above)
Comment (vibhcool) [inline src/org/loklak/api/search/TimeAndDateService.java]: fixed this

File: src/org/loklak/api/search/TimeAndDateService.java (hunk 2 above)
Comment (singhpratyush) [inline src/org/loklak/api/search/TimeAndDateService.java]: The sign thing is fixed. Can you look at the zero padding too?

Comment (vibhcool) [inline src/org/loklak/api/search/TimeAndDateService.java]: @singhpratyush added zero padding :)
